experiments/dsl/src/main/scala/org/apache/gearpump/streaming/flatmap/Stream.scala

```diff
@@ -0,0 +1,85 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.gearpump.streaming.flatmap
+
+import org.apache.gearpump.streaming.flatmap.op._
+import org.apache.gearpump.streaming.task.Task
+import org.apache.gearpump.util.Graph
+import org.slf4j.LoggerFactory
+
+class Stream[T](dag: Graph[Op, OpEdge], private val root: Op) {
+
+  def flatMap[R](fun: T => TraversableOnce[R]): Stream[R] = {
+    val flatMapOp = FlatMapOp(fun)
+    dag.addVertex(flatMapOp )
+    dag.addEdge(root, Direct, flatMapOp)
+    new Stream(dag, flatMapOp)
+  }
+
+  def map[R](fun: T => R): Stream[R] = {
+    this.flatMap { data =>
+      Option(fun(data))
+    }
+  }
+
+  def filter(fun: T => Boolean): Stream[T] = {
+    this.flatMap { data =>
+      if (fun(data)) Option(data) else None
+    }
+  }
+
+  def reduce(fun: (T, T) => T): Stream[T] = {
+    val reduceOp = ReduceOp(fun)
+    dag.addVertex(reduceOp)
+    dag.addEdge(root, Direct, reduceOp)
+    new Stream(dag, reduceOp)
+  }
+
+  def log(): Unit = {
+    this.map(msg => LoggerFactory.getLogger("dsl").info(msg.toString))
+
+  }
+
+  def merge(other: Stream[T]): Stream[T] = {
+    val mergeOp = MergeOp(root, other.root)
+    dag.addVertex(mergeOp)
+    dag.addEdge(root, Direct, mergeOp)
+    dag.addEdge(other.root, Shuffle, mergeOp)
+    new Stream(dag, mergeOp)
+  }
+
+  def groupBy[R](fun: T => R, parallism: Int): Stream[T] = {
```

Comment:
Rename R -> Group. 

class People(gender: String, name: String)
For example, for Stream[People]

We can group by gender, the output stream is still Stream[People]

Stream[People].groupBy(_.gender).reduce(...)